### PR TITLE
Getting rid of some micro optimizations in `ddm_irref`

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

I think that old sympy code used a lot of micro-optimizing, like assigning attributes to variable before loop
I'm testing out how such micro optimizations are worth it, by testing out `100 x 100` matrix
```
rng = Random(0)

def f():
    return Fraction(rng.randint(-10, 10), rng.randint(1, 10))

M = [[f() for _ in range(100)] for _ in range(100)]
%prun -s cumulative out = ddm_irref(M)
```
But I don't see there is any improvement even in 3.8, maybe because most of the times are dominated by `Fraction` and also

Especially after python 3.11, these kind of optimizations are likely be outdated because of specializing adapters
https://peps.python.org/pep-0659/
so things like `exquo = K.exquo` are going to be bad practice the future.

Maybe things like efficiently looping with both the key and the items, like `enumerate` or `dict.items` can still be valid in the future, and it wouldn't hurt the readability.
But generally, having too many variables, and abusing the variables make it hard to track down each state what they denote and how they mutate, so we need to get rid of something too trivial like `aij = a[i][j], aijinv = aij**-1, ...`

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
This is master:
```python
In [1]: for n in range(10):
   ...:     M = randMatrix(n)
   ...:     dM = M._rep.to_dense().to_field()
   ...:     %timeit dM.rref()
   ...: 
4.98 µs ± 165 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
7.74 µs ± 82.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
13 µs ± 1.98 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
19.5 µs ± 96.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
30.6 µs ± 90.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
53.8 µs ± 1.14 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
89.2 µs ± 4.08 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
124 µs ± 238 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
175 µs ± 1.89 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
240 µs ± 1.28 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
This is the PR:
```python
In [1]: for n in range(10):
   ...:     M = randMatrix(n)
   ...:     dM = M._rep.to_dense().to_field()
   ...:     %timeit dM.rref()
   ...: 
5.06 µs ± 171 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
7.57 µs ± 98.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
13.2 µs ± 678 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
24.2 µs ± 1.36 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
44.2 µs ± 4.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
74.5 µs ± 4.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
114 µs ± 5.07 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
174 µs ± 6.06 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
245 µs ± 6.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
364 µs ± 15.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by oscarbenjamin** [comment]
This is master:
```python
In [1]: M = randMatrix(100, seed=1)

In [2]: dM = M._rep.to_dense().to_field()

In [3]: %time ok = dM.rref()
CPU times: user 2.05 s, sys: 10.1 ms, total: 2.06 s
Wall time: 2.07 s
```
This is the PR:
```python
In [2]: M = randMatrix(100, seed=1)

In [3]: dM = M._rep.to_dense().to_field()

In [4]: %time ok = dM.rref()
CPU times: user 2.52 s, sys: 18.8 ms, total: 2.54 s
Wall time: 2.56 s
```

**Comment by oscarbenjamin** [comment]
The timings above are with CPython 3.8 on OSX.

**Comment by sylee957** [comment]
Okay, but I only keep the variable for the list and the result becomes negligible with 2 fewer variables

Master:
```
1.53 µs ± 26.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
2.34 µs ± 11.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
3.34 µs ± 21.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.37 µs ± 26.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
10.9 µs ± 75.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
17.9 µs ± 62.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
28.3 µs ± 50 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
43.2 µs ± 53.7 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
63.1 µs ± 154 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
87.8 µs ± 212 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
PR:
```
1.51 µs ± 5.31 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
2.28 µs ± 16 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
3.62 µs ± 22.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.13 µs ± 26.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
10.8 µs ± 20.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
18.1 µs ± 51.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
27.6 µs ± 35.7 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
41.8 µs ± 53.2 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
63.2 µs ± 537 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
89.8 µs ± 412 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

File: sympy/polys/matrices/dense.py [hunk 1]
```diff
@@ -116,38 +116,33 @@ def ddm_irref(a, _partial_pivot=False):
...
                 continue
 
         # normalise row
         ai = a[i]
-        aijinv = aij**-1
-        for l in range(j, n):
-            ai[l] *= aijinv # ai[j] = one
+        for l in range(j + 1, n):
+            ai[l] /= ai[j]
+        ai[j] /= ai[j]
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/dense.py:137]: Be very careful changing this and make sure to run timings with domains like `QQ(x, y)` or `QQ(sqrt(2))` etc. I'm not happy with exactly how this code works (#21834) but don't even think about changing this for reasons of aesthetics or micro-optimisation without a detailed investigation.

**Comment by oscarbenjamin** [comment]
> Okay, but I only keep the variable for the list and the result becomes negligible with 2 fewer variables

Maybe so but what positive improvement is achieved by the changes currently in this PR?

File: sympy/polys/matrices/dense.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/polys/matrices/dense.py:137]: Master:
```python
In [1]: from sympy import *
   ...: from sympy.polys.matrices import *
   ...: 
   ...: K = QQ.algebraic_field(sqrt(2))
   ...: for n in range(2, 10):
   ...:     M = Matrix(n, n, lambda i, j: (i**2+j+j*i)**10*sqrt(2) + j**2)
   ...:     dM = M._rep.convert_to(K)
   ...:     %time dM.inv()
   ...: 
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 787 µs
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 1.82 ms
CPU times: user 4 ms, sys: 0 ns, total: 4 ms
Wall time: 4.17 ms
CPU times: user 8 ms, sys: 0 ns, total: 8 ms
Wall time: 8.37 ms
CPU times: user 16 ms, sys: 0 ns, total: 16 ms
Wall time: 15.2 ms
CPU times: user 28 ms, sys: 0 ns, total: 28 ms
Wall time: 25.5 ms
CPU times: user 40 ms, sys: 0 ns, total: 40 ms
Wall time: 39.5 ms
CPU times: user 60 ms, sys: 0 ns, total: 60 ms
Wall time: 59.1 ms
```
PR:
```python
In [1]: from sympy import *
   ...: from sympy.polys.matrices import *
   ...: 
   ...: K = QQ.algebraic_field(sqrt(2))
   ...: for n in range(2, 10):
   ...:     M = Matrix(n, n, lambda i, j: (i**2+j+j*i)**10*sqrt(2) + j**2)
   ...:     dM = M._rep.convert_to(K)
   ...:     %time dM.inv()
   ...: 
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 1.36 ms
CPU times: user 4 ms, sys: 0 ns, total: 4 ms
Wall time: 3.29 ms
CPU times: user 8 ms, sys: 0 ns, total: 8 ms
Wall time: 6.86 ms
CPU times: user 12 ms, sys: 0 ns, total: 12 ms
Wall time: 12.7 ms
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 22 ms
CPU times: user 32 ms, sys: 0 ns, total: 32 ms
Wall time: 34.6 ms
CPU times: user 52 ms, sys: 0 ns, total: 52 ms
Wall time: 51.9 ms
CPU times: user 76 ms, sys: 0 ns, total: 76 ms
Wall time: 74.8 ms
```
Timings with CPython 3.8 on Linux.

**Comment by sylee957** [comment]
I assume things are loosely organized in `dense.py` I assumed we could use some fast functions for iterators in the bottom, than trying to unroll everything. 
Generally, these things are very mechanically done, and prone to mistakes.

**Comment by oscarbenjamin** [comment]
What do you mean by "fast functions for iterators"? Are you imagining something that is faster then just iterating over a list?

**Comment by sylee957** [comment]
I mean that ideally, we shouldn't redo things like having things like `transpose, zeros, ones` implemented somewhere and thrown out in the actual implementation because they don't operate in-place, make copy, or so on.

**Comment by oscarbenjamin** [comment]
When I originally wrote this code I did a lot of timings that resulted in the code being precisely the way that it is. I wanted to spend more time and do more timings but I didn't get round to doing that. If you want to improve this code then that is great but please make sure to take time to evaluate performance very carefully.

Right now I don't understand the purpose of this PR. The title is "getting rid of some micro optimisations in ddm_rref". Those micro optimisations were put there deliberately though because they improved speed at the time when I measured timings. These optimisations do not detract from readability or portability or anything else though. They are just slight tweaks to the code (changing division to multiplication is very different though - that is not a micro optimisation in this context).

Why would you want to remove these "optimisations"? What benefit do any of the changes in this PR give?

**Comment by oscarbenjamin** [comment]
> I mean that ideally, we shouldn't redo things like having things like `transpose, zeros, ones` implemented somewhere and thrown out in the actual implementation because they don't operate in-place, make copy, or so on.

Please be clearer about your goal. The changes in this PR bear no relation to your comment about transpose, zeros and ones.

**Comment by sylee957** [comment]
I'm not telling about the benefit, but I've just opened this to test some things out and to criticize some things inside

**Comment by sylee957** [comment]
That's why I set up this as draft

**Comment by oscarbenjamin** [comment]
To be clear I am in favour of improving this code. It just needs to be done very carefully. The intention of this particular code is that performance is the number one goal. Of course it can be better documented and potentially the code can be clearer but there the number one goal (performance) should not be sacrificed.

**Comment by oscarbenjamin** [comment]
> I've just opened this to test some things out and to criticize some things inside

Okay well please don't be put off from making improvements here. As I said I would like this code to be improved. It's fine to open a draft PR like this but the title of this PR made me think that you were intending to make aesthetic changes that would compromise the performance of this code but performance is more important than aesthetics here.

**Comment by asmeurer** [comment]
I'm generally in favor of not doing micro-optimizations in most places, but there are a few core "inner loop" functions that can be run thousands of times in a single operation and can be worth doing there (for instance, some places in the core or the low level stuff in the polys). It sounds like this particular function is also such an instance.

**Comment by sylee957** [comment]
I think that unfortunately, such micro optimizations should just be end up in vain if we encourage users to use fast libraries like `gmpy` or `flint` matrix for optional dependency. 
I especially optimizing tricks around pure python intrinsics become more dirty, and there is upper bound about how to achieve that.

**Comment by oscarbenjamin** [comment]
> I think that unfortunately, such micro optimizations should just be end up in vain if we encourage users to use fast libraries like `gmpy` or `flint` matrix for optional dependency.

The optimisations are not in vain for users who don't (or can't) install those optional dependencies.

> I especially find optimizing tricks around pure python intrinsics dirty

Please explain precisely what in this code you think is "dirty". The changes in this PR are cosmetic at best so it is not clear to me what problem you want to solve in general.

> there is upper bound about how much optimization we can achieve in pure python

I agree. We need to satisfy both cases though: fastest SymPy in pure Python without optional dependencies vs SymPy + flint etc.

**Comment by asmeurer** [comment]
Like I said, I generally agree with you for most cases. They just make the code harder to read and end up being destroyed if the code is changed even a little bit. But there are a few places in the code where the performance is the most important thing, even more important than these other considerations. There aren't many such places, and it would be good to keep them relatively isolated if possible. In these specific parts of the code, something that saves even 1% of the time can translate to seconds, because they are called thousands of times as part of a "top-level" operation like simplify() or solve() or integrate(). 

Based on https://github.com/sympy/sympy/pull/24580#issuecomment-1401011631, this is such an instance, because that's a difference of 20% for the final operation (and rref itself is often not the final operation, but part of the larger operation e.g. in integration).

**Comment by sylee957** [comment]
I think that the difference just comes from using not storing `aij**-1` as variable, 
Generally, the aesthetic code depends on the performance as well, because it removes redundancy
But in the core, I think that people are starting to use more mutation around states/variables to achieve that efficiency
and that makes the code specifically longer
That's the specific tricks what I refer by 'dirty'

**Comment by oscarbenjamin** [comment]
> They just make the code harder to read

Please give a concrete example with reference to this particular code. While I did micro optimise this I also wanted to keep the code clear and readable. I don't think I compromised readability in favour of performance (although I did compromise convenience, usability etc.).

> I think that the difference just comes from using not storing `aij**-1` as variable,

That is absolutely *not* a micro optimisation. Please look at how division works in algebraic number fields like `Q(sqrt(2))` and compare that with multiplication.

**Comment by sylee957** [comment]
I also referred to code like `aij = a[i][j]`, `exquo = K.exquo`
If they only give 1% improvements that are hardly to distinguish to noise, what are you going to do?

> Please look at how division works in algebraic number fields like Q(sqrt(2)) and compare that with multiplication.

I do agree that if computing inverse is involved, it can be stored as state.
I hardly recognize everything you had researched around defects about sympy,
because there is not even any sympy benchmarks pointing around them.

**Comment by oscarbenjamin** [comment]
> I also referred to code like `aij = a[i][j]`, `exquo = K.exquo`
> If they only give 1% improvements that are hardly to distinguish to noise, what are you going to do?

Nothing? Is `aij = a[i][j]` a problem in some way? Does it harm readability?

To be clear this is the function as it exists in master:
https://github.com/sympy/sympy/blob/a34e09ad26ef8da00c749aa47ee1497c82ecf6c7/sympy/polys/matrices/dense.py#L101-L160

> because there is not even any sympy benchmarks pointing about them for me to recognize.

This is a general problem. I have previously optimised things and then others have ruined that optimisation work for relatively unimportant reasons simply because they did not understand the performance implications. It is difficult to keep track of performance when multiple contributors edit the same code.

> But you need to be patient that I don't know everything you had researched around defects about sympy,

I think that I have overreacted here and I'm sorry for that. Please do try to improve this code.

**Comment by sylee957** [comment]
> Nothing? Is aij = a[i][j] a problem in some way? Does it harm readability?

You already understood `aij == a[i][j]` remain invariant, because you wrote it, 
but I need figure out why `aij` really points to same thing, that's what I refer that it is slightly more unreadable
because I need to do some math around operational semantics to see how it leads through.

But as I have noted, I am fine not doing any changes to it, 
and also fine doing any changes to it if is done in some umbrella of other nontrivial PR rally.
But I would approve if any other contributor do it, unless I don't find serious flaw or you jump in the reviews.

**Comment by oscarbenjamin** [comment]
> But I would approve if any other contributor do it, unless I don't find serious flaw or you jump in the reviews.

Please don't let any other contributor change this code unless serious analysis is done. This code can be improved but it is also very easy to cause problems when making changes here. Unless there is a seriously good reason no one should change anything at all.

These few lines of code are the workhorse behind solve, linsolve, heurisch etc. Many important potentially slow things go through this function. Improvements here can speed up many things but even a small change can cause many problems.

We could remove something like `aij = a[i][j]` but realistically I don't think that will make much difference to the readability of the code and no one should touch this code without a clear understanding of the (not very complicated) algorithm.

The thing that should really be improved here is the SDM version:
https://github.com/sympy/sympy/blob/a34e09ad26ef8da00c749aa47ee1497c82ecf6c7/sympy/polys/matrices/sdm.py#L1031-L1208
That is much more efficient for sparse cases like those that arise in heurisch but is not presently used by heurisch because of a subtle difference in pivoting.

**Comment by sylee957** [comment]
> That is much more efficient for sparse cases like those that arise in heurisch but is not presently used by heurisch because of a subtle difference in pivoting.

Unfortunately, that is very stressful objective because I don't see anything opaque about `heurisch` for which kind of subtlety that makes changes, or have identified at all.
For most higher level functions, the optimization should work like drop-in replacements, or if someone had messed it up, someone may need to restructure it.

**Comment by oscarbenjamin** [comment]
> For most higher level functions, the optimization should work like drop-in replacements, or if someone had messed it up, someone may need to restructure it.

Ideally it would be a drop-in replacement. Some analysis is needed to see why it is not. That analysis is much more valuable than fiddling around with trivial lines in `ddm_irref` (and it is better not to change anything in `ddm_irref` until the cause is understood).

**Comment by sylee957** [comment]
I'm just doing that in part time of the type analysis of sympy codebase, 
to clean up how sympy objects adapt to each other.

I hope you don't dismiss effort in that way,
and I'm neither getting paid or such, to have some responsibility to prioritize working on something people ask for.

**Pull request closed without merging**